File: src/main/java/eu/codetopic/utils/data/preferences/PreferencesData.java
package eu.codetopic.utils.data.preferences;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.SharedPreferences.OnSharedPreferenceChangeListener;
import android.os.Bundle;
import android.support.annotation.CallSuper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.LocalBroadcastManager;

import eu.codetopic.utils.PrefNames;
import eu.codetopic.utils.data.preferences.provider.SharedPreferencesProvider;
import eu.codetopic.utils.data.preferences.support.NoApplyPreferencesEditor;

public abstract class PreferencesData<SP extends SharedPreferences> {

    private static final String LOG_TAG = "PreferencesData";

    private static final String ACTION_DATA_CHANGED_BASE =
            "eu.codetopic.utils.data.preferences.PreferencesData.PREFERENCES_CHANGED.$1%s";
    public static final String EXTRA_CHANGED_DATA_KEY = "CHANGED_DATA_KEY";

    private final OnSharedPreferenceChangeListener mPreferenceChangeListener = (sharedPreferences, key) -> onChanged(key);

    private final Context mContext;
    private final SharedPreferencesProvider<SP> mPreferencesProvider;
    private final int mSaveVersion;

    private boolean mCreated = false;
    private boolean mDestroyed = false;

    public PreferencesData(Context context, @NonNull SharedPreferencesProvider<SP> preferencesProvider, int saveVersion) {
        mContext = context.getApplicationContext();
        mPreferencesProvider = preferencesProvider;
        mSaveVersion = saveVersion;
    }

    private static String getBroadcastActionChanged(@NonNull PreferencesData data) {
        String name = data.getName();
        return String.format(ACTION_DATA_CHANGED_BASE, (name == null ? "default" : name));
    }

    public final String getBroadcastActionChanged() {
        return getBroadcastActionChanged(this);
    }

    private Intent generateIntentActionChanged(String changedKey) {
        return new Intent(this.getBroadcastActionChanged())
                .putExtra(EXTRA_CHANGED_DATA_KEY, changedKey)
                .putExtras(getAdditionalDataChangedExtras(changedKey));
    }

    protected Bundle getAdditionalDataChangedExtras(String changedKey) {
        return new Bundle();
    }

    public Context getContext() {
        return mContext;
    }

    @Nullable
    public String getName() {
        return mPreferencesProvider.getName();
    }

    protected synchronized SharedPreferences getPreferences() {
        if (!mCreated) throw new IllegalStateException(LOG_TAG + " is not initialized");
        if (mDestroyed) throw new IllegalStateException(LOG_TAG + " is still destroyed");
        return mPreferencesProvider.getSharedPreferences();
    }

    protected synchronized SharedPreferences.Editor edit() {
        return getPreferences().edit();
    }

    public final synchronized void init() {
        if (mCreated) throw new IllegalStateException(LOG_TAG + " is still initialized");
        if (mDestroyed) throw new IllegalStateException(LOG_TAG + " is destroyed");
        mCreated = true;
        try {
            onCreate();
        } catch (Throwable t) {
            destroy();
            throw t;
        }
    }

    public final synchronized boolean isCreated() {
        return mCreated;
    }

    public final synchronized boolean isDestroyed() {
        return mDestroyed;
    }

    @CallSuper
    protected synchronized void onCreate() {
        getPreferences().registerOnSharedPreferenceChangeListener(mPreferenceChangeListener);
        checkUpgrade(mSaveVersion);
    }

    private synchronized void checkUpgrade(int saveVersion) {
        int actualVersion = getPreferences().getInt(PrefNames.DATA_SAVE_VERSION, -1);
        if (actualVersion != saveVersion) {
            SharedPreferences.Editor editor = edit();

            NoApplyPreferencesEditor noApplyEditor = new  NoApplyPreferencesEditor(editor,
                    "Don't call methods editor.apply() or editor.commit() " +
                            "during onUpgrade() or onDowngrade(). Changes will be saved later automatically.");
            if (saveVersion > actualVersion) onUpgrade(noApplyEditor, actualVersion, saveVersion);
            else onDowngrade(noApplyEditor, actualVersion, saveVersion);

            editor.putInt(PrefNames.DATA_SAVE_VERSION, saveVersion).apply();
        }
    }

    protected synchronized void onUpgrade(SharedPreferences.Editor editor, int from, int to) {
        // Default implementation will just throw all data away.
        editor.clear();
    }

    protected synchronized void onDowngrade(SharedPreferences.Editor editor, int from, int to) {
        // Default implementation should not support downgrading, so we will just throw exception.
        throw new UnsupportedOperationException("Version code cannot be downgraded");
    }

    @CallSuper
    protected synchronized void onChanged(String key) {
        LocalBroadcastManager.getInstance(mContext).sendBroadcast(generateIntentActionChanged(key));
    }

    @CallSuper
    protected synchronized void onDestroy() {
        getPreferences().unregisterOnSharedPreferenceChangeListener(mPreferenceChangeListener);
    }

    public final synchronized void destroy() {
        if (!mCreated) throw new IllegalStateException(LOG_TAG + " is not initialized");
        if (mDestroyed) throw new IllegalStateException(LOG_TAG + " is still destroyed");
        mDestroyed = true;
        mCreated = false;
        onDestroy();
    }

    @Override
    protected void finalize() throws Throwable {
        if (mCreated && !mDestroyed) destroy();
        super.finalize();
    }

    @Override
    public String toString() {
        return "PreferencesData{" +
                ", mSaveVersion=" + mSaveVersion +
                ", mPreferencesProvider=" + mPreferencesProvider +
                ", mCreated=" + mCreated +
                ", mDestroyed=" + mDestroyed +
                '}';
    }
}

File: src/main/java/eu/codetopic/utils/data/preferences/PreferencesData.kt
/*
 * utils
 * Copyright (C)   2017  anty
 *
 * This program is free  software: you can redistribute it and/or modify
 * it under the terms  of the GNU General Public License as published by
 * the Free Software  Foundation, either version 3 of the License, or
 * (at your option) any  later version.
 *
 * This program is distributed in the hope that it  will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied  warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You  should have received a copy of the GNU General Public License
 * along  with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package eu.codetopic.utils.data.preferences

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.os.Bundle
import android.support.annotation.CallSuper
import eu.codetopic.utils.broadcast.LocalBroadcast
import eu.codetopic.utils.data.preferences.provider.ISharedPreferencesProvider
import eu.codetopic.utils.edit

abstract class PreferencesData<out SP : SharedPreferences> (
        context: Context, private val preferencesProvider: ISharedPreferencesProvider<SP>) :
        IPreferencesData {

    companion object {

        private const val LOG_TAG = "PreferencesData"

        const val EXTRA_CHANGED_DATA_KEY = "CHANGED_DATA_KEY"

        private fun getBroadcastActionChanged(data: PreferencesData<*>): String =
                "eu.codetopic.utils.data.preferences.PreferencesData" +
                        ".PREFERENCES_CHANGED.${data.name ?: "default"}"
    }

    private val preferenceChangeListener = OnSharedPreferenceChangeListener { _, key -> onChanged(key) }

    protected val context: Context = context.applicationContext

    protected val accessProvider: ISharedPreferencesProvider<SP>
            by lazy { PreferencesProvider() }

    final override var isCreated = false
        private set

    final override var isDestroyed = false
        private set

    final override val broadcastActionChanged: String
        get() = getBroadcastActionChanged(this)

    final override val name: String?
        get() = preferencesProvider.name

    protected val preferences: SP
        get() {
            if (!isCreated) throw IllegalStateException("$LOG_TAG is not initialized")
            if (isDestroyed) throw IllegalStateException("$LOG_TAG is still destroyed")
            return preferencesProvider.preferences
        }

    private fun generateIntentActionChanged(changedKey: String?): Intent {
        return Intent(this.broadcastActionChanged)
                .putExtra(EXTRA_CHANGED_DATA_KEY, changedKey)
                .putExtras(getAdditionalDataChangedExtras(changedKey))
    }

    protected open fun getAdditionalDataChangedExtras(changedKey: String?): Bundle {
        return Bundle()
    }

    @Suppress("DeprecatedCallableAddReplaceWith")
    @Deprecated("Migrate to Kotlin and use new edit(block) instead.")
    @SuppressLint("CommitPrefEdits")
    protected fun edit(): SharedPreferences.Editor = preferences.edit()

    protected inline fun edit(block: SharedPreferences.Editor.() -> Unit) = preferences.edit(block)

    @Synchronized
    final override fun init() {
        if (isCreated) throw IllegalStateException("$LOG_TAG is still initialized")
        if (isDestroyed) throw IllegalStateException("$LOG_TAG is destroyed")
        isCreated = true
        try {
            onCreate()
        } catch (t: Throwable) {
            try {
                destroy()
            } catch (_: Throwable) {}
            throw t
        }

    }

    @CallSuper
    @Synchronized
    protected open fun onCreate() {
        //Log.v(LOG_TAG, "onCreate(name=$name)")
        preferences.registerOnSharedPreferenceChangeListener(preferenceChangeListener)
    }

    @CallSuper
    @Synchronized
    protected open fun onChanged(key: String?) {
        //Log.v(LOG_TAG, "onChanged(name=$name, key=$key)")
        LocalBroadcast.sendBroadcast(generateIntentActionChanged(key))
    }

    @CallSuper
    @Synchronized
    protected open fun onDestroy() {
        //Log.v(LOG_TAG, "onDestroy(name=$name)")
        preferences.unregisterOnSharedPreferenceChangeListener(preferenceChangeListener)
    }

    @Synchronized
    final override fun destroy() {
        if (!isCreated) throw IllegalStateException("$LOG_TAG is not initialized")
        if (isDestroyed) throw IllegalStateException("$LOG_TAG is still destroyed")
        onDestroy()
        isDestroyed = true
        isCreated = false
    }

    protected open fun finalize() {
        if (isCreated && !isDestroyed) destroy()
    }

    override fun toString(): String {
        return "PreferencesData(preferencesProvider=$preferencesProvider, " +
                "isCreated=$isCreated, isDestroyed=$isDestroyed)"
    }

    protected inner class PreferencesProvider : ISharedPreferencesProvider<SP> {

        override val name: String?
            get() = this@PreferencesData.name

        override val preferences: SP
            get() = this@PreferencesData.preferences

    }
}
